### ISSUE TITLE
Could not read property java.math.BigDecimal 

### ISSUE THREAD

**Issue by hevan** [jasync-sql/jasync-sql#113]
{"timestamp":1553798370728,"path":"/customer/find/2","status":500,"error":"Internal Server Error","message":"Could not read property java.math.BigDecimal com.xp.ws.busi.domain.Customer.amount from result set!"}

**Comment by hevan** [comment]
Mysql 8.0 GA

**Comment by hevan** [comment]
@Table("customer")
@Data
@AllArgsConstructor
@NoArgsConstructor
public class Customer {
	
	@Id String id;
	
	String name;
	
	@Transient
	@DateTimeFormat(pattern = "yyyy-MM-dd")
	@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd", timezone = "GMT+8")
	LocalDate createdate;
	
	@DateTimeFormat(pattern = "yyyy-MM-dd")
	@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd", timezone = "GMT+8")
	LocalDate updatedate;

	BigDecimal amount;
	
	String content;
}

**Comment by oshai** [comment]
Can you please try to use jasync 0.9.38?

**Comment by oshai** [comment]
Also, what is the type of amount in the db?

**Comment by hevan** [comment]
2019-03-29 03:07:51.667 ERROR 31095 --- [onPool-worker-3] a.w.r.e.AbstractErrorWebExceptionHandler : [646b0089] 500 Server Error for HTTP GET "/customer/find/2"

org.springframework.data.mapping.MappingException: Could not read property java.math.BigDecimal com.xp.ws.busi.domain.Customer.amount from result set!
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.readFrom(MappingR2dbcConverter.java:144) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:117) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:101) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:59) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:46) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:29) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at com.github.jasync.r2dbc.mysql.JaysncResult$map$1.apply(JaysncResult.kt:25) ~[jasync-r2dbc-mysql-0.9.38.jar!/:na]
        at com.github.jasync.r2dbc.mysql.JaysncResult$map$1.apply(JaysncResult.kt:13) ~[jasync-r2dbc-mysql-0.9.38.jar!/:na]
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:107) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath(FluxIterable.java:243) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:201) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxFlatMap$FlatMapInner.onSubscribe(FluxFlatMap.java:933) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:139) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:63) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxMapFuseable.subscribe(FluxMapFuseable.java:63) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.Flux.subscribe(Flux.java:7777) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:389) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.innerNext(FluxConcatMap.java:275) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.FluxConcatMap$ConcatMapInner.onNext(FluxConcatMap.java:849) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1505) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at reactor.core.publisher.MonoCompletionStage.lambda$subscribe$0(MonoCompletionStage.java:82) ~[reactor-core-3.2.6.RELEASE.jar!/:3.2.6.RELEASE]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962) ~[na:1.8.0_181]
        at com.github.jasync.sql.db.util.FutureUtilsKt.success(FutureUtils.kt:17) ~[jasync-common-0.9.38.jar!/:na]
        at com.github.jasync.sql.db.mysql.MySQLConnection$succeedQueryPromise$1.accept(MySQLConnection.kt:255) ~[jasync-mysql-0.9.38.jar!/:na]
        at com.github.jasync.sql.db.mysql.MySQLConnection$succeedQueryPromise$1.accept(MySQLConnection.kt:48) ~[jasync-mysql-0.9.38.jar!/:na]
        at java.util.Optional.ifPresent(Optional.java:159) ~[na:1.8.0_181]
        at com.github.jasync.sql.db.mysql.MySQLConnection.succeedQueryPromise(MySQLConnection.kt:254) ~[jasync-mysql-0.9.38.jar!/:na]
        at com.github.jasync.sql.db.mysql.MySQLConnection.onResultSet(MySQLConnection.kt:264) ~[jasync-mysql-0.9.38.jar!/:na]
        at com.github.jasync.sql.db.mysql.codec.MySQLConnectionHandler.handleEOF(MySQLConnectionHandler.kt:435) ~[jasync-mysql-0.9.38.jar!/:na]
        at com.github.jasync.sql.db.mysql.codec.MySQLConnectionHandler.channelRead0(MySQLConnectionHandler.kt:126) ~[jasync-mysql-0.9.38.jar!/:na]
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[netty-transport-4.1.33.Final.jar!/:4.1.33.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.33.Final.jar!/:4.1.33.Final]

**Comment by hevan** [comment]
mysql> desc customer;
+------------+---------------+------+-----+---------+-------+
| Field      | Type          | Null | Key | Default | Extra |
+------------+---------------+------+-----+---------+-------+
| id         | varchar(50)   | YES  |     | NULL    |       |
| name       | varchar(100)  | YES  |     | NULL    |       |
| updatedate | datetime      | YES  |     | NULL    |       |
| content    | longtext      | YES  |     | NULL    |       |
| amount     | decimal(16,2) | YES  |     | NULL    |       |
+------------+---------------+------+-----+---------+-------+
5 rows in set (0.00 sec)

**Comment by oshai** [comment]
I think I was able to reproduce and fix it. can you please try jasync version 0.9.39 ?

**Comment by hevan** [comment]
use jasync version 0.9.39  the same error

**Comment by hevan** [comment]
org.springframework.data.mapping.MappingException: Could not read property java.math.BigDecimal com.xp.ws.busi.domain.Customer.amount from result set!
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.readFrom(MappingR2dbcConverter.java:144) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:117) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:101) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:59) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:46) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]
        at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:29) ~[spring-data-r2dbc-1.0.0.gh-75-SNAPSHOT.jar!/:1.0.0.gh-75-SNAPSHOT]

**Comment by hevan** [comment]
Can use this  Mysql library， for product APP， in future， Have   risk？

**Comment by oshai** [comment]
jasync is ready for production, but the reactive spring support is not yet ready (ie gh-75 branch). regardless, let me try again to fix it.

**Comment by oshai** [comment]
I managed to reproduce it, but it looks more like a spring issue.
The complete stacktrace I get is:
```
reactor.core.Exceptions$ErrorCallbackNotImplemented: org.springframework.data.mapping.MappingException: Could not read property @org.springframework.data.relational.core.mapping.Column(value=amount, keyColumn=) @javax.validation.constraints.Digits(message={javax.validation.constraints.Digits.message}, groups=[], payload=[], integer=16, fraction=2)private java.math.BigDecimal com.example.rx.Order.amount from result set!
Caused by: org.springframework.data.mapping.MappingException: Could not read property @org.springframework.data.relational.core.mapping.Column(value=amount, keyColumn=) @javax.validation.constraints.Digits(message={javax.validation.constraints.Digits.message}, groups=[], payload=[], integer=16, fraction=2)private java.math.BigDecimal com.example.rx.Order.amount from result set!
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.readFrom(MappingR2dbcConverter.java:144) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:117) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:101) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.read(MappingR2dbcConverter.java:59) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:46) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.EntityRowMapper.apply(EntityRowMapper.java:29) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	at com.github.jasync.r2dbc.mysql.JasyncResult$map$1.apply(JasyncResult.kt:25) ~[jasync-r2dbc-mysql-0.9.39.jar:na]
	at com.github.jasync.r2dbc.mysql.JasyncResult$map$1.apply(JasyncResult.kt:13) ~[jasync-r2dbc-mysql-0.9.39.jar:na]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:107) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath(FluxIterable.java:243) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:201) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxFlatMap$FlatMapInner.onSubscribe(FluxFlatMap.java:933) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:139) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:63) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxMapFuseable.subscribe(FluxMapFuseable.java:63) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.Flux.subscribe(Flux.java:7777) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:389) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1505) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at reactor.core.publisher.MonoCompletionStage.lambda$subscribe$0(MonoCompletionStage.java:82) ~[reactor-core-3.2.6.RELEASE.jar:3.2.6.RELEASE]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760) ~[na:1.8.0_192]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736) ~[na:1.8.0_192]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[na:1.8.0_192]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962) ~[na:1.8.0_192]
	at com.github.jasync.sql.db.util.FutureUtilsKt.success(FutureUtils.kt:17) ~[jasync-common-0.9.39.jar:na]
	at com.github.jasync.sql.db.mysql.MySQLConnection$succeedQueryPromise$1.accept(MySQLConnection.kt:255) ~[jasync-mysql-0.9.39.jar:na]
	at com.github.jasync.sql.db.mysql.MySQLConnection$succeedQueryPromise$1.accept(MySQLConnection.kt:48) ~[jasync-mysql-0.9.39.jar:na]
	at java.util.Optional.ifPresent(Optional.java:159) ~[na:1.8.0_192]
	at com.github.jasync.sql.db.mysql.MySQLConnection.succeedQueryPromise(MySQLConnection.kt:254) ~[jasync-mysql-0.9.39.jar:na]
	at com.github.jasync.sql.db.mysql.MySQLConnection.onResultSet(MySQLConnection.kt:264) ~[jasync-mysql-0.9.39.jar:na]
	at com.github.jasync.sql.db.mysql.codec.MySQLConnectionHandler.handleEOF(MySQLConnectionHandler.kt:435) ~[jasync-mysql-0.9.39.jar:na]
	at com.github.jasync.sql.db.mysql.codec.MySQLConnectionHandler.channelRead0(MySQLConnectionHandler.kt:126) ~[jasync-mysql-0.9.39.jar:na]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) ~[netty-codec-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) ~[netty-codec-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_192]
Caused by: java.lang.IllegalStateException: Required identifier property not found for class java.math.BigDecimal!
	at org.springframework.data.mapping.PersistentEntity.getRequiredIdProperty(PersistentEntity.java:105) ~[spring-data-commons-2.2.0.M2.jar:2.2.0.M2]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.readEntityFrom(MappingR2dbcConverter.java:182) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	at org.springframework.data.r2dbc.function.convert.MappingR2dbcConverter.readFrom(MappingR2dbcConverter.java:137) ~[spring-data-r2dbc-1.0.0.gh-75-20190320.145400-1.jar:1.0.0.gh-75-SNAPSHOT]
	... 53 common frames omitted
```

When debugging it, it looks like spring think BigDecimal is a complex entity, or at least that is what I figured out. @mp911de - can you please take a look?

**Comment by mp911de** [comment]
This is a Spring Data issue. Please file a ticket in https://github.com/spring-projects/spring-data-r2dbc and attach the last stack trace. `BigDecimal` is considered to be an entity instead of a simple (primitive) type and that's why reading fails.

I think you can close this ticket here @oshai.

**Comment by oshai** [comment]
Thanks, I am closing it. @hevan please file the issue in https://github.com/spring-projects/spring-data-r2dbc

**Comment by sysmat** [comment]
this is basic type and this driver not support it, it is a shame
